libdnf/rpm/solv_query.cpp

```diff
@@ -37,16 +37,89 @@ extern "C" {
 
 #include <fnmatch.h>
 
+namespace {
+
+inline bool is_valid_candidate(libdnf::sack::QueryCmp cmp_type, const char * c_pattern, const char * candidate, bool all_candidates) {
+    switch (cmp_type) {
+        case libdnf::sack::QueryCmp::EQ: {
+            return strcmp(candidate, c_pattern) == 0;
+        } break;
+        case libdnf::sack::QueryCmp::IEXACT: {
+            return strcasecmp(candidate, c_pattern) == 0;
+        } break;
+        case libdnf::sack::QueryCmp::GLOB: {
+            return all_candidates || fnmatch(c_pattern, candidate, 0) == 0;
+        } break;
+        case libdnf::sack::QueryCmp::IGLOB: {
+            return all_candidates || fnmatch(c_pattern, candidate, FNM_CASEFOLD) == 0;
+        } break;
+        default:
+            throw libdnf::rpm::SolvQuery::NotSupportedCmpType("Unsupported CmpType");
+    }
+}
+
+bool is_valid_candidate(Pool * pool, libdnf::rpm::PackageId candidate_id, Id src, bool all_epoch, bool all_version, bool all_release, bool all_arch, bool test_epoch, bool test_version, bool test_release, bool test_arch, const char * epoch_c_pattern, const char * version_c_pattern, const char * release_c_pattern, const char * arch_c_pattern, libdnf::sack::QueryCmp epoch_cmp_type, libdnf::sack::QueryCmp version_cmp_type, libdnf::sack::QueryCmp release_cmp_type, libdnf::sack::QueryCmp arch_cmp_type) {
+    if (src != 0) {
+        if (libdnf::rpm::solv::get_solvable(pool, candidate_id)->arch == src) {
+            return false;
+        }
+    }
+    if (test_arch) {
+        auto candidate_arch = libdnf::rpm::solv::get_arch(pool, candidate_id);
+        if (!is_valid_candidate(arch_cmp_type, arch_c_pattern, candidate_arch, all_arch)) {
+            return false;
+        }
+    }
+    if (test_epoch) {
+        auto candidate_epoch = libdnf::rpm::solv::get_epoch_cstring(pool, candidate_id);
+        if (!is_valid_candidate(epoch_cmp_type, epoch_c_pattern, candidate_epoch, all_epoch)) {
+            return false;
+        }
+    }
+    if (test_version) {
+        auto candidate_version = libdnf::rpm::solv::get_version(pool, candidate_id);
+        if (!is_valid_candidate(version_cmp_type, version_c_pattern, candidate_version, all_version)) {
+            return false;
+        }
+    }
+    if (test_release) {
+        auto candidate_release = libdnf::rpm::solv::get_release(pool, candidate_id);
+        if (!is_valid_candidate(release_cmp_type, release_c_pattern, candidate_release, all_release)) {
+            return false;
+        }
+    }
+    return true;
+}
+
+inline bool hy_is_glob_pattern(const char * pattern) {
```

Comment:
Remove the `hy_` prefix.
Move to libdnf::utils as an internal, header-only inline function.